client/src/stream.rs

```diff
@@ -152,6 +193,20 @@ impl rpc::Server for CallbackServer {
                     Ok(CallbackResp::DeviceChange)
                 })
             }
+            CallbackReq::SharedMem(mut handle, shm_area_size) => {
+                let shm = unsafe {
+                    SharedMem::from(handle.take_handle(), shm_area_size)
+                        .expect("Client failed to set up shmem")
+                };
+                self.shm = Some(shm);
+
+                self.duplex_input = if let StreamDirection::Duplex = self.dir {
+                    Some(Vec::with_capacity(shm_area_size))
```

Comment:
How do we know `shm_area_size` is larger than `nframes as usize * input_frame_size`?

Reply:
Unfortunately we don't have a way to precisely determine the required shmem size right now, so we choose a sufficiently large over-estimate instead (which will be specified by a Gecko pref).  The ideal solution to this is to extend cubeb's API with a way to query a configured stream's required buffer size before starting it - that will also require auditing each backend to ensure the reported buffer size is fixed (some backends may grow the buffer in response to latency, for instance).  I filed PR https://github.com/mozilla/audioipc-2/issues/124 to track this.

Reply:
Answered here: https://github.com/mozilla/audioipc-2/pull/121#discussion_r683090394